### PR TITLE
Create tests to fix #473

### LINKED ISSUE #473: Import/Export to external database fails for PostgreSQL in 3.0 of jabref
Hi,

Just tried 3.0.  The import and export features for PostgreSQL fail with the following messages:

For Export - "Could not export to SQL database for the following reason:
                     This connection has been closed"

For Import - "Could not import from SQL database for the following reason:
                      This ResultSet is closed."

Establishing a connection to the database returns successfully.

Thanks

Fred

### PR DESCRIPTION
**Pull request by simonharrer**
- [x] Test which reproduces #473
- [x] fix for #473

### CODE REVIEW

**Comment by simonharrer** [comment]
Possible solution: use prepared statements, which should solve this issue automatically. And they should be faster and safer to execute as well.

**Comment by Siedlerchr** [comment]
@simonharrer  Yep, I would support that. And second you could use them for MYSQL as well

**Comment by simonharrer** [comment]
Another issue: the we currently support the export of a JabRef database to an SQL file using the syntax for MySQL or PostgreSQL. But this SQL file does not contain the correct escape logic, and it is very hard to get this info out of JDBC without a database connection. Hence, I would suggest to remove these actions, as I cannot see how they are useful without the correct escape logic.

File: src/main/java/net/sf/jabref/sql/exporter/DatabaseExporter.java [hunk 1]
```diff
@@ -170,36 +155,38 @@ private int populateEntryGroupsTable(GroupTreeNode cursor, int parentID, int cur
...
-                    currentID = populateEntryGroupsTable(child, myID, currentID, out, database_id);
+                    currentID = populateEntryGroupsTable(child, myID, currentID, connection, database_id);
                 }
-                //Unfortunatley, AutoCloseable throws only Exception
-            } catch (Exception e) {
-                LOGGER.warn("Cannot close resource", e);
-            }
 
+            }
+            //Unfortunatley, AutoCloseable throws only Exception
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/sql/exporter/DatabaseExporter.java]: Remove comment

File: src/main/java/net/sf/jabref/sql/exporter/DatabaseExporter.java [hunk 2]
```diff
@@ -305,28 +290,25 @@ private int populateGroupsTable(GroupTreeNode cursor, int parentID, int currentI
...
+                ResultSet rs = statement.executeQuery(
+                        "SELECT groups_id FROM groups WHERE label='" + cursor.getGroup().getName() + "' AND database_id='"
+                                + database_id + "' AND parent_id='" + parentID + "';")) {
+            // setting values to ID and myID to be used in case of textual SQL
+            // export
+            int myID = currentID;
+            rs.next();
+            myID = rs.getInt("groups_id");
+            for (GroupTreeNode child : cursor.getChildren()) {
+                ++currentID;
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/sql/exporter/DatabaseExporter.java]: currentID++;

File: src/main/java/net/sf/jabref/sql/exporter/DatabaseExporter.java [hunk 3]
```diff
@@ -305,28 +290,25 @@ private int populateGroupsTable(GroupTreeNode cursor, int parentID, int currentI
...
+                currentID = populateGroupsTable(child, myID, currentID, out, database_id);
             }
-
+            //Unfortunatley, AutoCloseable throws only Exception
```
Comment (stefan-kolb) [inline src/main/java/net/sf/jabref/sql/exporter/DatabaseExporter.java]: You used SQLException before?!

**Comment by stefan-kolb** [comment]
LGTM :+1:

File: src/main/java/net/sf/jabref/sql/exporter/DatabaseExporter.java (hunk 3 above)
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/sql/exporter/DatabaseExporter.java]: Maybe take a look at what I did in #1265
I replaced all throws Exception with their typed ones

**Pull request merged**